Use checksummed input streams in `validate_checksums()`

ndragazis opened the pull request:
With commits ed7d352e7d and bb1867c7c7, we now have input streams for both compressed and uncompressed SSTables that provide seamless checksum and digest checking. The code for these was based on `validate_checksums()`, which implements its own validation logic over raw streams. This has led to some duplicate code.

This PR deduplicates the uncompressed case by modifying `validate_checksums()` to use a checksummed input stream instead of a raw stream. The same cannot be done for compressed SSTables though. The reason is that `validate_checksums()` needs to examine the whole data file, even if an invalid chunk is encountered. In the checksummed case we support that by offloading the error handling logic from the data source via a function parameter. In the compressed data source we cannot do that because it needs to return decompressed data and decompression may fail if the data are invalid.

This PR also enables `validate_checksums()` to partially verify SSTables with just the per-chunk checksums if the digest is missing. 

In more detail, this PR consists of:
* Port of some integrity checks from `do_validate_uncompressed()` to the checksummed data source. It should now be able to detect corruption due to truncated or appended chunks (expected number of chunks is retrieved from the CRC component).
* Introduction of `error_handler` parameter in checksummed data source and `data_stream()`.
* Refactoring of `validate_checksums()`. The JSON response of `sstable validate-checksums` was also modified to report a missing digest.
*  Tests for `validate_checksums()` against SSTables with truncated data, appended data, invalid digests, or no digest.

Refs #19058.

This PR is a hybrid of cleanup and feature. No backport is needed.

scylladb-promoter commented:
## Docs Preview :book:
Docs Preview for this pull request is available [here](http://previews.docs.scylladb.com/scylladb/20933)

**Changed Files:**
- [operating-scylla/admin-tools/scylla-sstable](http://previews.docs.scylladb.com/scylladb/20933/operating-scylla/admin-tools/scylla-sstable)

**Note:** This preview will be available for 30 days and will be automatically deleted after that period. You can manually trigger a new build by committing changes.

ndragazis commented:
Following the merge of #20720, I have rebased and undrafted this PR.

avikivity commented:
> Thanks to PRs https://github.com/scylladb/scylladb/pull/20207 and https://github.com/scylladb/scylladb/pull/20720 

Please quote commits, not pull requests, commits are accessible via `git log` and git tools while pull requests require using github.

avikivity commented:
What functionality does this add?

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12467/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12467/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: boost/sstable_datafile_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12467/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1178/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4244/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12270/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12467/)
#### Build Details:
- Duration: 4 hr 23 min
- Builder: spider4.cloudius-systems.com

denesb commented on sstables/checksummed_data_source.cc:
```diff
@@ -17,6 +17,7 @@
 #include "exceptions.hh"
 #include "checksum_utils.hh"
 #include "data_source_types.hh"
+#include "checksummed_data_source.hh"
```
How did we get away without having this include so far?

ndragazis replied:
What's the problem here? Definition without declaration? I don't see `-Wmissing-prototypes` in flags.

denesb replied:
No problem per se, I'm just wondering how the code compiled without this header was included. Seems like a missing include. Maybe I'm missing something. Anyway, I was just thinking out loud.

ndragazis replied:
Prior to this PR, the header provided the declarations of the two factory methods, so it was only used by the consumers of the API.

With this PR, it also contains the type definition for the error handler, which is necessary in the function definitions as well.

denesb commented on sstables/checksummed_data_source.hh:
```diff
@@ -15,15 +15,19 @@
 
 namespace sstables {
 
+using integrity_error_handler_t = std::function<void(seastar::sstring)>;
```
`integrity_error_handler` is a good enough name, the `_t` suffix is not needed (we don't really use it).

denesb replied:
You should also declare here (and define in the `.cc` file) a default error handler:

```
integrity_error_handler throwing_integrity_error_handler;
```

So that users have a ready-to-use one for the default case. Maybe even make this the default argument for `error_handler`. We usually don't like default arguments but in this case we have a default used by almost all the callers and we have one exceptional user who will use a different one.

ndragazis replied:
> `integrity_error_handler` is a good enough name, the `_t` suffix is not needed (we don't really use it).

Ack, we do use it in some places, but in general we are not consistent.
```
$ git grep 'using [a-zA-Z0-9_]* = std::function' | wc -l
78
$ git grep 'using [a-zA-Z0-9_]*_t = std::function' | wc -l
12
$ git grep 'using [a-zA-Z0-9_]*_type = std::function' | wc -l
14
$ git grep 'using [a-zA-Z0-9_]*_func = std::function' | wc -l
14
$ git grep 'using [a-zA-Z0-9_]*_fn = std::function' | wc -l
11
```

denesb commented on sstables/sstables.hh:
```diff
@@ -730,10 +732,18 @@ public:
+            raw_stream raw = raw_stream::no, integrity_check integrity = integrity_check::no,
+            std::optional<integrity_error_handler_t> error_handler = [] (sstring msg) {
+                throw sstables::malformed_sstable_exception(msg);
+            });
```
The optional combined with the default parameter doesn't make sense. Either keep the optional and default to `std::nullopt`, or keep the default but make this a non-optional.

ndragazis commented:
>What functionality does this add?

1. First two patches move some checks from `validate_checksums()` to the checksummed data source. This makes them available for scrub as well.
2. Lets `sstable validate-checksums` proceed with checksum validation, even if the digest is missing. Also, it adds a boolean key in the JSON response to indicate if the digest is missing, and renames key `valid_checksums` to `valid`. With all that, the new response will be:
    ```
    {
         "has_checksums": Bool,
         "has_digest": Bool,
         "valid": Bool, // optional
     }
    ```

avikivity commented:
> > What functionality does this add?
> 
> 1. First two patches move some checks from `validate_checksums()` to the checksummed data source. This makes them available for scrub as well.
> 2. Lets `sstable validate-checksums` proceed with checksum validation, even if the digest is missing. Also, it adds a boolean key in the JSON response to indicate if the digest is missing, and renames key `valid_checksums` to `valid`. With all that, the new response will be:
>    ```
>    {
>         "has_checksums": Bool,
>         "has_digest": Bool,
>         "valid": Bool, // optional
>     }
>    ```


Ok. Makes sense. Is there risk of a performance regression, or is this just moving things around?

ndragazis commented:
The only change in the critical path is a check for premature EOF in the checksummed data source.
patch: https://github.com/scylladb/scylladb/pull/20933/commits/226bab486b8be4491aa63963f061a86b7257fbd8

ndragazis commented:
v2 Changelog:

Resolved review comments:
* `s/integrity_error_handler_t/integrity_error_handler`
* Defined `throwing_integrity_error_handler`, used it as default argument in `make_checksummed_{m, k_l}_format_input_stream()`.
* Made `error_handler` parameter in `data_stream()` non-optional.

Tests:
* Added test for truncated SSTables.
   We already had one, but it would truncate the data file after loading the SSTable, so the cached file size would be the old one. The new one performs a pre-load truncation, which is more realistic. I didn't remove the old one.
* Added test for SSTables with random appended data.

Chunk count mismatch:
* Removed patch "sstables: Compare number of chunks in CRC and Data"
   which would cause the checksummed data source to fail early in the constructor if there were a chunk count mismatch. We want `validate_checksums()` to examine the checksums and the digest regardless. This was the behavior before and it shouldn't change.
* Added patch "sstables: Check for excessive chunks in checksummed data source".
* Patch "sstables: Refactor validate_checksums() to use checksummed data stream":
   Changed the check in `do_validate_uncompressed()`. Instead of checking the data length, check the number of chunks. We know the expected number of chunks from the CRC component, but we don't know the exact data length because the last chunk may be smaller than the chunk size.
* Patch "sstables: Check for premature EOF in checksummed data source":
   Same change: use chunks instead of the file length.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12952/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12952/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: boost/sstable_datafile_test
#### Failed Tests (1/148):
- [test_validate_checksums](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12952/testReport/junit/boost/sstable_datafile_test/test_validate_checksums) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_validate_checksums)
#### Build Details:
- Duration: 1 hr 56 min
- Builder: i-06973ccd17bda8db5 (m5ad.8xlarge)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13105/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13105/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: boost/sstable_datafile_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/179/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4719/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12748/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1644/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13105/)
#### Failed Tests (2/37138):
- [test_random_failures[stop_after_setting_mode_to_normal_raft_topology-restart_coordinator_node].debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13105/testReport/junit/%28root%29/test_random_failures/Tests___Unit_Tests___test_random_failures_stop_after_setting_mode_to_normal_raft_topology_restart_coordinator_node__debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_setting_mode_to_normal_raft_topology-restart_coordinator_node].debug.2)
- [test_case_sensitive_type_name](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13105/testReport/junit/user_types_test/TestUserTypes/Tests___dtest_with_topology_changes___test_case_sensitive_type_name) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_case_sensitive_type_name)
#### Build Details:
- Duration: 8 hr 37 min
- Builder: i-0f7ba47e43cf04661 (m5ad.8xlarge)

ndragazis commented:
v3 Changelog:
* Fixed failing unit test by making a small change on the premature EOF condition.
   
   The test was failing with an SSTable that had its **last** chunk truncated **after load**. In this case, calling `get()` after reading the last chunk from the stream would throw a `runtime_error` because the position would not be aligned at the chunk boundary.
   
   This is a very extreme case that affects only `validate-checksums` (requires not throwing after an invalid checksum), and requires a concurrent truncation of the SSTable's last chunk while the tool is running. I'm submitting a fix just because we already had a test for this case. The fix (essentially the `else-if` statement):
    ```c
    -            if ((buf.size() & ~(chunk_size - 1)) == 0 && chunk_index + 1 < _checksum.checksums.size()) {
    -                throw malformed_sstable_exception(seastar::format("Checksummed reader hit premature end-of-file at file offset {}: expected {} chunks of size {} but data file has {}",
    -                        _underlying_pos + buf.size(), _checksum.checksums.size(), chunk_size, chunk_index + 1));
    +            if ((buf.size() & ~(chunk_size - 1)) == 0) {
    +                auto actual_end = _underlying_pos + buf.size();
    +                if (chunk_index + 1 < _checksum.checksums.size()) {
    +                    throw malformed_sstable_exception(seastar::format("Checksummed reader hit premature end-of-file at file offset {}: expected {} chunks of size {} but data file has {}",
    +                            actual_end, _checksum.checksums.size(), chunk_size, chunk_index + 1));
    +                } else if (actual_end < _file_len) {
    +                    // Truncation on last chunk. Update _end_pos so that future
    +                    // calls to get() return immediately.
    +                    _end_pos = actual_end;
    +                }
    ```
    (Change incorporated in patch: "sstables: Check for premature EOF in checksummed data source")                                   
* Added test to trigger this bug more easily by stripping just one byte from the last chunk. In contrast, the test from v2 strips half of the data.
   (Change incorporated in patch: "test: test_validate_checksums: Complement test for truncated SSTable")

denesb commented:
@ndragazis conflicts, please rebase.

mykaul commented:
> @ndragazis conflicts, please rebase.

@ndragazis - don't. Please hold off working on this until we finish higher priority items, such as tablets issues.
CC @denesb , @bhalevy for awareness.

denesb commented:
> > @ndragazis conflicts, please rebase.
> 
> @ndragazis - don't. Please hold off working on this until we finish higher priority items, such as tablets issues. CC @denesb , @bhalevy for awareness.

This PR just needs a rebase and is otherwise good to go in. Why let it bitrot for months, to avoid a rebase which takes 10 minutes?

mykaul commented:
> > > @ndragazis conflicts, please rebase.
> > 
> > 
> > @ndragazis - don't. Please hold off working on this until we finish higher priority items, such as tablets issues. CC @denesb , @bhalevy for awareness.
> 
> This PR just needs a rebase and is otherwise good to go in. Why let it bitrot for months, to avoid a rebase which takes 10 minutes?

10 minutes + additional reviewer's work.
I don't mind the 10 minutes, I mind the rest.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13349/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13349/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: boost/sstable_datafile_test
:x: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/352/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1859/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12972/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13349/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4966/)
#### Build Details:
- Duration: 5 hr 20 min
- Builder: spider7.cloudius-systems.com

denesb commented:
> ❌ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/352/)

There is a problem with this test, will re-kick CI after it is fixed.
